doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,229 @@
+The logic for determining what timestamp to return must satisfy:
+
+- If it's a read, then it is larger than or equal to the `since` of all objects involved in the query.
+- If it's a write, then it is larger than or equal to the `upper` of the table.
+- It is larger than or equal to the global timestamp
+- It is larger than or equal to the timestamp of the most recently completed query.
+    - If the query is a write and the most recently completed query was a read, then the timestamp returned must be
+      strictly larger than the previous timestamp.
+
+If some operation is given timestamp `ts`, then once that operation completes the global timestamp must be larger than
+or equal to `ts`. How this specific property is achieved for reads is described
+in [Read Capabilities on Global Timestamp](#Read Capabilities on Global Timestamp) and how it's achieved for writes is
+described in [Group Commit/Write Coalesce](#Group Commit/Write Coalesce).
+
+The properties described in this section are sufficient to ensure Strict Serializability across one start of
+Materialize (i.e. not between restarts).
+
+- [X] All reads and writes occur at specific timestamps: satisfied by the Coordinator assigning timestamps to all
+  operations.
+- [X] These timestamps are non-decreasing in real-time order: all timestamps are assigned sequentially on a single
+  thread. Each timestamp is guaranteed to be larger than or equal to the timestamp of the most recently completed query.
+- [X] With strict increases for writes following reads: the global timestamp explicitly increases after all writes (
+  See [Group Commit/Write Coalesce](#Group Commit/Write Coalesce)).
+- [X] Each timestamp is assigned to an event at some real-time moment within the event's bounds: For any two events e1
+  and e2, if e2 starts after e1 finished, then e2's timestamp will be larger or equal timestamp than e1's timestamp.
+- [X] Each timestamp transition is made durable before any subsequent response is issued: writes wait for an `append`
+  command to be processed before returning a response to the client.
+    - NOTE: There is ongoing work to ensure that an `append` command doesn't return until it is durable.
+
+### Global Timestamp Recovery
+
+After a restart the Coordinator must reestablish the global timestamp to some value greater than or equal to the value
+of the previous global timestamp before the restart. This ensures that a read or write after the restart is assigned a
+timestamp greater than or equal to all reads and writes before the restart.
+
+For an initial implementation the Coordinator will record all global timestamps in the catalog. On restart the
+Coordinator can read this value from the catalog.
+
+TODO: better implementation.
+
+The properties described in this section and the [Global Timestamp](#Global Timestamp) section are sufficient to ensure
+Strict Serializability across restarts.
+
+### Group Commit/Write Coalesce
```

Comment:
One thing I just thought of is, can the writes that belong to an `UPDATE` be committed at the same timestamp as the writes that belong to an `INSERT` to the same table? I'm not sure, but I'll think about it. @sploiselle you may have some thoughts since you were looking at `ReadThenWrite` related things.

Reply:
Yes! We are able to order those however we want, and ReadThenWrite only needs to guarantee that the read set hasn't changed. So if the write time is shared, that's fine because it would not have changed the read set.

Reply:
Awesome, thanks for the clarification!